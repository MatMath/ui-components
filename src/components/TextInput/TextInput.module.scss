@use '../../colors/colors';
@use '../../typography/typography';

/* remove edge decorators */
::-ms-clear {
  display: none;
}

.root {
  display: flex;
  flex-direction: column;
  align-items: flex-start;

  .input {
    /* typography */
    @include typography.body-medium;

    /* colors */
    color: colors.$black;
    background: colors.$white;

    /* geometry */
    padding: 0 16px;
    height: 40px;
    outline: none;
    display: flex;
    align-items: center;

    /* border */
    border: 1px solid colors.$dash-green-03;
    border-radius: 4px;
    box-sizing: border-box;

    &:focus {
      border-color: colors.$mid-green-00;
    }

    &.warning,
    &.input:focus.warning {
      border-color: colors.$orange-00;
    }

    &.error,
    &.input:focus.error {
      border-color: colors.$functional-red-02;
    }

    &:disabled {
      background: colors.$dash-green-06;
      border-color: colors.$dash-green-05;
      color: colors.$dash-green-03;
    }

    &:read-only {
      background: colors.$dash-green-06;
      border: 1px solid colors.$dash-green-05;
    }

    &::placeholder {
      color: colors.$dash-green-03;
    }

    &::-ms-input-placeholder {
      color: colors.$dash-green-03;
    }
  }

  .feedback {
    @include typography.body-small;
    color: colors.$black;
    margin-top: 4px;
  }

  .feedback.warning {
    color: colors.$orange-00;
  }

  .feedback.error {
    color: colors.$functional-red-02;
  }
}
