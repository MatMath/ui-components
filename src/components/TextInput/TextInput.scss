@use '../../colors/colors';
@use '../../typography/typography';

.input {
  color: colors.$black;
}

.root {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}

.input {
  /* typography */
  @include typography.body-medium;

  /* colors */
  color: colors.$black;
  background: colors.$white;

  /* geometry */
  padding: 0 16px;
  height: 40px;
  outline: none;
  display: flex;
  align-items: center;

  /* border */
  border: 1px solid colors.$dash-green-03;
  border-radius: 4px;
  box-sizing: border-box;
}

.input:focus {
  border: 1px solid colors.$mid-green-00;
}

.input.warning,
.input:focus.warning {
  border: 1px solid colors.$orange-00;
}

.input.error,
input:focus.error {
  border: 1px solid colors.$functional-red-02;
}

.input:disabled {
  background: colors.$dash-green-06;
  border: 1px solid colors.$dash-green-05;
  color: colors.$dash-green-03;
}

.input:read-only {
  background: colors.$dash-green-06;
  border: 1px solid colors.$dash-green-05;
}

.input::placeholder {
  color: colors.$dash-green-03;
}

.input::-ms-input-placeholder {
  color: colors.$dash-green-03;
}

/* remove edge decorators */
::-ms-clear {
  display: none;
}

.feedback {
  @include typography.body-small;
  color: colors.$black;
  margin-top: 4px;
}

.feedback.warning {
  color: colors.$orange-00;
}

.feedback.error {
  color: colors.$functional-red-02;
}
