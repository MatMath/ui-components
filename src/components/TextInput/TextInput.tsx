import * as React from 'react';
import style from './TextInput.module.scss';

type TextInputFeedbackType = 'error' | 'warning';

type _ForbiddenProps = 'size' | 'prefix' | 'type';

interface TextInputProps
  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, _ForbiddenProps> {
  placeholder?: string;
  feedbackType?: TextInputFeedbackType;
  feedbackText?: string;
}

const classNameFromFeedbackType: Record<TextInputFeedbackType, string> = {
  error: 'error',
  warning: 'warning'
};

const getClassNameFromFeedbackType = (feedbackType?: TextInputFeedbackType) => {
  return feedbackType ? classNameFromFeedbackType[feedbackType] : '';
};

const shouldShowFeedback = ({ feedbackText, feedbackType }: TextInputProps) => {
  return feedbackText && feedbackType;
};

export const TextInput: React.FC<TextInputProps> = props => {
  const className = getClassNameFromFeedbackType(props.feedbackType);
  const styledClassName = style[className] ? style[className] : '';

  return (
    <div className={`${style.root} ${styledClassName}`}>
      <input
        {...props}
        className={`${style.input} ${styledClassName}`}
        type='text'
      />
      {shouldShowFeedback(props) ? (
        <span className={`${style.feedback} ${styledClassName}`}>
          {props.feedbackText}
        </span>
      ) : null}
    </div>
  );
};
