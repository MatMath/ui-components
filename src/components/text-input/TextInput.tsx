import * as React from 'react';
import styles from './TextInput.module.scss';

type TextInputFeedbackType = 'error' | 'warning';

type _ForbiddenProps = 'size' | 'prefix' | 'type';

interface TextInputProps
  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, _ForbiddenProps> {
  placeholder?: string;
  feedbackType?: TextInputFeedbackType;
  feedbackText?: string;
}

const shouldShowFeedback = ({ feedbackText, feedbackType }: TextInputProps) => {
  return feedbackText && feedbackType;
};

export const TextInput: React.FC<TextInputProps> = props => {
  const styledClassName =
    props.feedbackType && styles[props.feedbackType]
      ? styles[props.feedbackType]
      : '';

  return (
    <div className={`${styles.root} ${styledClassName}`}>
      <input
        {...props}
        className={`${styles.input} ${styledClassName}`}
        type='text'
      />
      {shouldShowFeedback(props) ? (
        <span className={`${styles.feedback} ${styledClassName}`}>
          {props.feedbackText}
        </span>
      ) : null}
    </div>
  );
};
