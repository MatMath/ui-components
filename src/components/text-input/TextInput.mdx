---
name: TextInput
menu: Components
---

import { Playground, Props } from 'docz';
import { TextInput } from './TextInput';
import '../../utility/playgroundStyles.css';

# TextInput

A text input component

## Properties

<Props of={TextInput} />

## Usage

### Basic usage

Examples below show usage of the TextInput

<Playground className={{ preview: 'playground-base' }}>
  <TextInput />
</Playground>
