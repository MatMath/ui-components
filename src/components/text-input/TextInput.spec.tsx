import * as React from 'react';
import { mount } from 'enzyme';
import 'jest-styled-components';
import { TextInput } from './TextInput';

describe('<TextInput>', () => {
  describe('Global render', () => {
    it('should render default', () => {
      const textInput = mount(<TextInput />);
      expect(textInput).toMatchSnapshot();
    });

    it('should render placeholder', () => {
      const textInput = mount(<TextInput placeholder='placeholder' />);

      expect(textInput).toMatchSnapshot();
    });

    it('should render error', () => {
      const textInput = mount(<TextInput feedbackType='error' />);

      expect(textInput).toMatchSnapshot();
    });

    it('should render warning', () => {
      const textInput = mount(<TextInput feedbackType='warning' />);

      expect(textInput).toMatchSnapshot();
    });

    it('should render disabled', () => {
      const textInput = mount(<TextInput disabled />);

      expect(textInput).toMatchSnapshot();
    });

    it('should render warning and show feedback text', () => {
      const textInput = mount(
        <TextInput feedbackType='warning' feedbackText='warning' />
      );

      expect(textInput).toMatchSnapshot();
    });

    it('should render with fluid class', () => {
      const textInput = mount(<TextInput fullWidth />);

      expect(textInput).toMatchSnapshot();
      expect(textInput.props().fullWidth).toEqual(true);
      expect(textInput.find('input').getElement().props).toHaveProperty(
        'className',
        'input fullWidth'
      );
    });
  });
});
