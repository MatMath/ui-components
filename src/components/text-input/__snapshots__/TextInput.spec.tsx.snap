// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput> Global render should render default 1`] = `
<Component>
  <div
    className="root"
  >
    <input
      className="input"
      type="text"
    />
  </div>
</Component>
`;

exports[`<TextInput> Global render should render disabled 1`] = `
<Component
  disabled={true}
>
  <div
    className="root"
  >
    <input
      className="input"
      disabled={true}
      type="text"
    />
  </div>
</Component>
`;

exports[`<TextInput> Global render should render error 1`] = `
<Component
  feedbackType="error"
>
  <div
    className="root error"
  >
    <input
      className="input error"
      type="text"
    />
  </div>
</Component>
`;

exports[`<TextInput> Global render should render placeholder 1`] = `
<Component
  placeholder="placeholder"
>
  <div
    className="root"
  >
    <input
      className="input"
      placeholder="placeholder"
      type="text"
    />
  </div>
</Component>
`;

exports[`<TextInput> Global render should render warning 1`] = `
<Component
  feedbackType="warning"
>
  <div
    className="root warning"
  >
    <input
      className="input warning"
      type="text"
    />
  </div>
</Component>
`;

exports[`<TextInput> Global render should render warning and show feedback text 1`] = `
<Component
  feedbackText="warning"
  feedbackType="warning"
>
  <div
    className="root warning"
  >
    <input
      className="input warning"
      type="text"
    />
    <span
      className="feedback warning"
    >
      warning
    </span>
  </div>
</Component>
`;
